datumaro/plugins/data_formats/synthia/base.py

```diff
@@ -96,78 +72,89 @@ def parse_label_map(path):
-            gt_images = find_images(inst_dir, recursive=True)
-            for gt_img in gt_images:
-                item_id = osp.splitext(osp.relpath(gt_img, inst_dir))[0].replace("\\", "/")
+        if self._inst_dir and osp.isdir(self._inst_dir):
+            gt_labels = glob(self._inst_dir + "/*.txt")
+            for gt_label in gt_labels:
+                item_id = osp.splitext(osp.relpath(gt_label, self._inst_dir))[0].replace("\\", "/")
 
                 anno = []
-                labels_mask = load_image(gt_img, dtype=np.uint16)
-                dynamic_objects = np.unique(labels_mask[:, :, 1])
-                labels_mask = labels_mask[:, :, 2]
-                segm_ids = np.unique(labels_mask)
-                for segm_id in segm_ids:
-                    attr = {"dynamic_object": False}
-                    if segm_id != 0 and segm_id in dynamic_objects:
-                        attr["dynamic_object"] = True
+                labels_mask = np.loadtxt(gt_label)
+                classes = np.unique(labels_mask)
+                for label_id in classes:
                     anno.append(
-                        Mask(
-                            image=self._lazy_extract_mask(labels_mask, segm_id),
-                            label=segm_id,
-                            attributes=attr,
-                        )
+                        Mask(image=self._lazy_extract_mask(labels_mask, label_id), label=label_id)
                     )
 
                 image = images.get(item_id)
                 if image:
                     image = Image.from_file(path=image)
 
                 items[item_id] = DatasetItem(id=item_id, media=image, annotations=anno)
+        elif self._seg_dir and osp.isdir(self._seg_dir):
+            for seg_img_path in find_images(self._seg_dir, recursive=True):
+                item_id = osp.splitext(osp.relpath(seg_img_path, self._seg_dir))[0].replace(
+                    "\\", "/"
+                )
 
-        elif osp.isdir(osp.join(root_dir, SynthiaPath.SEMANTIC_SEGM_DIR)):
-            gt_dir = osp.join(root_dir, SynthiaPath.SEMANTIC_SEGM_DIR)
-            gt_images = find_images(gt_dir, recursive=True)
-            for gt_img in gt_images:
-                item_id = osp.splitext(osp.relpath(gt_img, gt_dir))[0].replace("\\", "/")
-
-                anno = []
                 inverse_cls_colormap = self._categories[AnnotationType.mask].inverse_colormap
-                color_mask = lazy_mask(gt_img, inverse_cls_colormap)
-                color_mask = color_mask()
+
+                seg_img = load_image(seg_img_path, dtype=np.uint16)
+                color_mask = np.zeros_like(seg_img[:, :, 0])
+                for i in range(seg_img.shape[0]):
+                    for j in range(seg_img.shape[1]):
+                        pixel_color = tuple(seg_img[i, j, :])
+                        if pixel_color in inverse_cls_colormap.keys():
+                            color_mask[i, j] = inverse_cls_colormap[pixel_color]
```

Comment:
Basically, I think it is not good idea to loop over (H,W) dimensions with Python because of performance issues. On the other hand, this code will dump all masks in memory (not lazy loading). The original code probably wasn't lazy loading as well (calling `color_mask()`). However, we can use the previous logic to achieve fully lazy loading as follows.

```python
lazy_color_mask = lazy_mask(gt_img, inverse_cls_colormap)
lazy_color_mask_callable = lambda: lazy_color_mask() == label_id

...

for label_id in classes:
    anno.append(
          Mask(image=lazy_color_mask_callable, label=label_id)
    )
```

Reply:
I updated at the [241b5548af138265e96f7c871eec0b522a225448](https://github.com/openvinotoolkit/datumaro/pull/987/commits/241b5548af138265e96f7c871eec0b522a225448)

Reply:
Is there any pixel from the segmentation masks of Synthia which cannot be mapped with the table? I guess that Synthia leaves this `("Void", (0, 0, 0))` mapping for `background` or `ignored`. Isn't it our mistake that we can't find the mapping rule for those pixels?

Reply:
I left another comment to compare the predefined label map and undeclared pixels in the Synthia-SF dataset.
Please take a look at below.